[clang][LoongArch] Don't imply -fdirect-access-external-data for non-PIC

heiher opened the pull request:
LoongArch does not and will not support copy relocation, so non-PIC code defaults to -fno-direct-access-external-data.

Explicit `-fdirect-access-external-data` is acceptable, for example if the code is writable.

Reported-by: Xi Ruoyao <xry111@xry111.site>
Signed-off-by: WANG Rui <wangrui@loongson.cn>

Fixes #71645

heiher commented:
Add @xry111 @xen0n

xen0n reviewed (commented):
Looks okay, although I may be missing some places. For now I don't know if recording this fact (no copy reloc support) with a helper function somewhere would help de-duplicating the logic -- it should be the case normally but I don't know where to best put the helper.

xry111 commented:
Suggest to add "Fixes #71645" in the message.

xry111 commented:
cc @MaskRay

heiher committed:
[clang][LoongArch] Don't imply -fdirect-access-external-data for non-PIC

LoongArch does not and will not support copy relocation, so non-PIC code
defaults to -fno-direct-access-external-data.

Explicit `-fdirect-access-external-data` is acceptable, for example if
the code is writable.

Reported-by: Xi Ruoyao <xry111@xry111.site>
Signed-off-by: WANG Rui <wangrui@loongson.cn>

Fixes #71645

MaskRay commented:
Thanks for the patch, but I think we can keep the default only in the default so that Frontend does not need to handle target-specific differences. See 
https://github.com/llvm/llvm-project/pull/72221 and I've added only few tests that make sense (`-fpic -fdirect-access-external-data` generally doesn't make sense. Newer tests should use `--target=`. We should also avoid relying on the -fno-pic default for baremetal.)
